### PR TITLE
Made Mish Activation function support complex dtype

### PR DESCRIPTION
**Pull request by mohame54**
# PR Description 
Made some modification to **mish** backend implementations and their corresponding tests 
For complex dtype support

### CODE REVIEW

**Comment by ivy-leaves** [comment]
If you are working on an open task, please **edit** the PR description to link to the issue you've created.

For more information, please check [ToDo List Issues Guide](https://unify.ai/docs/ivy/overview/contributing/the_basics.html#todo-list-issues).

Thank you :hugs:

File: ivy/functional/backends/paddle/elementwise.py [hunk 1]
```diff
@@ -719,9 +719,12 @@ def tanh(
...
     if paddle.is_complex(x):
-        tanh_a = paddle.tanh(paddle.real(x))
-        tan_b = paddle.tan(paddle.imag(x))
-        return (tanh_a + 1j * tan_b) / (1 + 1j * (tanh_a * tan_b))
+        const = paddle.to_tensor(1j, dtype=x.dtype)
+        tanh_a = paddle.cast(paddle.tanh(x.real()), x.dtype)
+        tan_b = paddle.cast(paddle.tan(x.imag()), x.dtype)
+        term1 = tanh_a + paddle.multiply(tan_b, const)
+        term2 = 1 + paddle.multiply(const, (tanh_a * tan_b))
+        return term1 / term2
```
Comment (jshepherd01) [inline ivy/functional/backends/paddle/elementwise.py:727]: I'm not sure I see the reason for changing this.

Also, rather than casting a float to complex and then multiplying by 1j, it would probably be better to use the `paddle.complex(real, imag)` function as in:
```py
tanh_a = paddle.tanh(paddle.real(x))
tan_b = paddle.tan(paddle.imag(x))
term1 = paddle.complex(tanh_a, tan_b)
term2 = paddle.complex(1, tanh_a * tan_b)
return term1 / term2
```

File: ivy_tests/test_ivy/test_functional/test_nn/test_activations.py [hunk 2]
```diff
@@ -136,14 +136,19 @@ def test_log_softmax(*, dtype_and_x, axis, test_flags, backend_fw, fn_name, on_d
...
+        min_dim_size=2,
         safety_factor_scale="log",
     ),
+    ground_truth_backend="jax",
 )
 def test_mish(*, dtype_and_x, test_flags, backend_fw, fn_name, on_device):
     dtype, x = dtype_and_x
+    x = x[0]
+    if len(x.shape) < 2:
+        x = x.reshape(-1, 1)
```
Comment (jshepherd01) [inline ivy_tests/test_ivy/test_functional/test_nn/test_activations.py:151]: Why is this needed? Calling `ivy.mish` on a 1D or even 0D array shouldn't cause any problems

**Review by jshepherd01 — COMMENTED** [review]
Looks good to me, just a couple of questions to clarify some things

**Comment by jshepherd01** [comment]
You might also want to consider increasing the safety factor even further, it seems to fail tests for any reasonably large value of x (which is reasonable, because it needs to compute numbers on the order of `e^{2*Re[x]}`, so an x value of about 350 would be very near to an overflow for complex128)

File: ivy/functional/backends/paddle/elementwise.py (hunk 1 above)
Comment (mohame54) [reply ivy/functional/backends/paddle/elementwise.py:727]: I have done that cuz paddle was giving me errors related to paddle functional API but I will modify the tanh implementation  as yo suggested.

File: ivy_tests/test_ivy/test_functional/test_nn/test_activations.py (hunk 2 above)
Comment (mohame54) [reply ivy_tests/test_ivy/test_functional/test_nn/test_activations.py:151]: it's something related also to paddle backend when I extract the real and imag part of the complex input it transforms them to **0D** although the input is 1D dimension array.

**Comment by mohame54** [comment]
okay on it.

File: ivy/functional/backends/paddle/elementwise.py (hunk 1 above)
Comment (mohame54) [reply ivy/functional/backends/paddle/elementwise.py:727]: it seems there is a problem with paddle.complex not creating output if the real and imaginary part not of the same type
so I kept it the same before .
image

Comment (jshepherd01) [reply ivy/functional/backends/paddle/elementwise.py:727]: It's likely the 1, you might need to do `paddle.ones_like(tanh_a)`

File: ivy_tests/test_ivy/test_functional/test_nn/test_activations.py (hunk 2 above)
Comment (jshepherd01) [reply ivy_tests/test_ivy/test_functional/test_nn/test_activations.py:151]: Isn't this still doing `x[0]` afterwards, so if the input was 1D then it's still getting a 1D array but only with one element

Comment (mohame54) [reply ivy_tests/test_ivy/test_functional/test_nn/test_activations.py:151]: x is initially a list of arrays I took the first array from the list and expand it's dims.

File: ivy/stateful/module.py [hunk 3]
```diff
@@ -6,7 +6,7 @@
 import os
 import abc
 import copy
-import dill
+#import dill
```
Comment (jshepherd01) [inline ivy/stateful/module.py:9]: `dill` seems to be used in this file, you should leave the import in

File: ivy_tests/test_ivy/test_functional/test_nn/test_activations.py (hunk 2 above)
Comment (jshepherd01) [reply ivy_tests/test_ivy/test_functional/test_nn/test_activations.py:151]: Yes, but the test still does `x[0]` afterwards, so for an input of say `x = [ivy.array([1,2,3])]` this process turns it into `x = ivy.array([[1], [2], [3]])` and then the input to `test_function` is `x = x[0] = ivy.array([1])`. I also tried removing this section and re-running it locally and the tests still passed

File: ivy/functional/backends/torch/activations.py [hunk 4]
```diff
@@ -125,13 +125,16 @@ def log_softmax(
 @with_unsupported_dtypes(
     {
         "2.0.1 and below": (
-            "complex",
+            "bfloat16",
```
Comment (jshepherd01) [inline ivy/functional/backends/torch/activations.py:128]: This surprises me, torch is usually fine with `bfloat16` on cpu. Are you sure it's the function that doesn't support it, rather than the testing pipeline?

**Review by jshepherd01 — CHANGES_REQUESTED** [review]

**Comment by mohame54** [comment]
>Yes, but the test still does x[0] afterwards, so for an input of say x = [ivy.array([1,2,3])] this process turns it into x = ivy.array([[1],
>[2], [3]]) and then the input to test_function is x = x[0] = ivy.array([1]). I also tried removing this section and re-running it locally >and the tests still passed
Oh, I totally forgot about it.

**Comment by jshepherd01** [comment]
You might need to run the precommit again, the linter is still failing.

Also about the `x.reshape` thing, I really think that shouldn't be necessary in the test since `ivy.mish(1+2j)` should, in principle, work. If removing that causes tests to fail, it should probably be fixed in the backend functions themselves rather than by modifying the test

**Comment by mohame54** [comment]
it causes errors with paddle backend only

**Comment by jshepherd01** [comment]
> it causes errors with paddle backend only

I haven't been able to reproduce the error, would you be able to provide a screenshot of it?

**Comment by mohame54** [comment]
>I haven't been able to reproduce the error, would you be able to provide a screenshot of it?

No worries managed to solve them.

File: ivy/functional/ivy/activations.py [hunk 5]
```diff
@@ -663,16 +663,27 @@ def softsign(
...
+    *,
+    complex_mode: Literal["split", "magnitude", "jax"] = "jax",
+    out: Optional[ivy.Array] = None,
 ) -> ivy.Array:
     """
     Apply the mish activation function element-wise.
 
+    If the input is complex we apply the mish function to the input
+    ``mish(x) = x * tanh(log(1+ e^x))``.
+
```
Comment (jshepherd01) [inline ivy/functional/ivy/activations.py:679]: I think you'd be fine without this part of the docstring, since that's also what happens when the input is real

**Review by jshepherd01 — CHANGES_REQUESTED** [review]
Just a minor docstring change and also the thing about the function signatures that I mentioned on the logit PR applies here too

**Comment by mohame54** [comment]
>Just a minor docstring change and also the thing about the function signatures that I mentioned on the logit PR applies here too
yeah I totally forgot about them.

**Comment by jshepherd01** [comment]
> > Just a minor docstring change and also the thing about the function signatures that I mentioned on the logit PR applies here too
> yeah I totally forgot about them.

No worries, I keep forgetting about them too
